### DISCUSSION TITLE
급하고 초보입니다ㅠㅠ)카운트에 따라 오브젝트 없어짐

### DISCUSSION THREAD

**Discussion by maskpop** [naverz/zepeto-studio-kor#713] [5. Q&A]
안녕하세요 천천히 알아가는 재미에 제작중인 학생입니다.

찾아도 나오지 않는 부분에 대해 질문드립니다... 제가 잘 알고있는지도 모르겠어서 막막하더라구요

오브젝트에 캐릭터가 닿으면 오브젝트가 Destroy되고 카운트가 되는데, 3이 되었을 때 포탈을 막고있는 구조물 또한 Destroy 되도록 만들고 싶습니다.

유니티 상에서는 에러가 뜨지 않는데 실제 실행해보면 오브젝트는 부서지는데 구조물은 없어지지가 않더라구요... 도움 부탁드립니다...!

export default class mission extends ZepetoScriptBehaviour {

   public Potalopener:GameObject;
   public Missonblock:GameObject;
   
   private num:number

    Start()
    {
    
    }

    OnTriggerEnter(obj:Collider) //트리거에 들어가면
    {
       this.Missionb(obj.gameObject) //미션b 실행
    }

    Missionb(obj:GameObject) //미션b는
    {
        this.num++; //숫자가 늘어나고
        GameObject.Destroy(this.Missonblock); //블럭이부서진다
    }

    Update() 
    {    
        if(this.num == 3) //숫자가 3이 되었을때
        {
            GameObject.Destroy(this.Potalopener); //포탈오프너가부서진다
        }

    }

}

**Comment by x4cf** [comment]
스타트에서 초기화를 한번 해보시겠어요?
private num:number => private num: int;

Start()
{
  this.num = 0;
}

**Reply by maskpop** [reply]
바꾸어 봐도 부서지는 것만 되고 구조물이 없어지지는 않았어요 ㅠㅠ 그리고 세개의 오브젝트가 부서져야하는데 이런식으로 만들면 public에 cube 하나만 있어서 하나만 넣을 수 있게 되더라구요...

![캡처](https://user-images.githubusercontent.com/105036393/167433042-d816520f-9209-43e6-b1ae-bf516b6a722a.PNG)

**Comment by smilejsu82** [comment]
간단하게 만들어 봤습니다 

https://user-images.githubusercontent.com/50188795/167341093-42f68c4d-fac5-48c8-b00f-9b3bef18f40e.mp4

![image](https://user-images.githubusercontent.com/50188795/167341176-b2d6aa11-e94e-40ce-9ce3-530991a6c5b0.png)


`import { ZepetoScriptBehaviour } from 'ZEPETO.Script'
import { GameObject, Object } from 'UnityEngine'
import Cube from './Cube'
export default class App extends ZepetoScriptBehaviour {

    public  barricadeGo : GameObject;
    public cubes : GameObject[];
    private count : number = 0;
    
    Start() {    

        this.cubes.forEach((go, idx)=>{
            const cube = go.GetComponent<Cube>();
            cube.distroyEvent.AddListener(()=>{
                this.count++; 
                console.log(`count : ${this.count}`);
                cube.gameObject.SetActive(false);
                if(this.count >= 3){
                     this.barricadeGo.SetActive(false);
                }
            });
        });
    }

}`

`import { ZepetoScriptBehaviour } from "ZEPETO.Script";
import { UnityEvent } from 'UnityEngine.Events'
import { Collider } from 'UnityEngine'
export default class Cube extends ZepetoScriptBehaviour {
  
  public distroyEvent : UnityEvent= new UnityEvent();
  
  OnTriggerEnter(col : Collider){
    if(col.tag == "Player")
    {
      this.distroyEvent.Invoke();  
    }
  }
}
`
`import { ZepetoScriptBehaviour } from "ZEPETO.Script";
import {
  SpawnInfo,
  ZepetoPlayers,
  LocalPlayer,
  ZepetoCharacter,
} from "ZEPETO.Character.Controller";

export default class CharacterControllerSample extends ZepetoScriptBehaviour {
  Start() {
    ZepetoPlayers.instance.CreatePlayerWithZepetoId(
      "",
      "smilejsu82",
      new SpawnInfo(),
      true
    );

    ZepetoPlayers.instance.OnAddedLocalPlayer.AddListener(() => {
      let _player: LocalPlayer = ZepetoPlayers.instance.LocalPlayer;
      _player.tag = "Player";
    });
  }
}
`

**Reply by maskpop** [reply]
그리고 new Event를 추가하는 이유는 뭔지 알 수 있을까요...?

**Reply by maskpop** [reply]
똑같이 만들었는데도 큐브가 사라지지 않아요... Cube라는 스크립트는 따로 어딘가에 넣어야 하나요??

**Reply by smilejsu82** [reply]
> 그리고 new Event를 추가하는 이유는 뭔지 알 수 있을까요...?

이벤트를 사용하기 위함 입니다 ~ 
가이드를 참고해보세요 ~ 
https://studio.zepeto.me/guides/events

**Reply by smilejsu82** [reply]
> 똑같이 만들었는데도 큐브가 사라지지 않아요... Cube라는 스크립트는 따로 어딘가에 넣어야 하나요??

Cube 스크립트는 Cube 게임 오브젝트에 넣어 주세요~

**Reply by maskpop** [reply]
Cube 스크립트를 닿으면 없어져야 하는 큐브게임 오브젝트에 넣고 시도해보고 있습니다! 아래대로 넣었는데 똑같이 사람이  큐브 통과만 됩니다 ㅠㅠ
![image](https://user-images.githubusercontent.com/105036393/167521717-89d26d40-b547-48a0-8a3a-fdbcb2d696be.png)

![image](https://user-images.githubusercontent.com/105036393/167521881-fa764be3-836b-4741-ace5-a13977f1d9ce.png)

**Reply by smilejsu82** [reply]
> Cube 스크립트를 닿으면 없어져야 하는 큐브게임 오브젝트에 넣고 시도해보고 있습니다! 아래대로 넣었는데 똑같이 사람이  큐브 통과만 됩니다 ㅠㅠ
> 
> ![image](https://user-images.githubusercontent.com/105036393/167521717-89d26d40-b547-48a0-8a3a-fdbcb2d696be.png)
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/105036393/167521881-fa764be3-836b-4741-ace5-a13977f1d9ce.png)
> 
> 
> 
> 

플레이어에 태그가 Player인지 확인좀 부탁해요~
그리구 각각 큐브에 리지드바디 부착후 use gravity는 체크 해제 해주시구여

**Reply by maskpop** [reply]
큐브에 리지드바디 부착 후 use gravity 해제했습니다! 그랬는데도 똑같네요... 플레이어 태그라는게 이부분이 맞나요??

![image](https://user-images.githubusercontent.com/105036393/167526987-badbdcb5-6bf4-4596-865f-89029e4b323c.png)

**Reply by smilejsu82** [reply]
> 큐브에 리지드바디 부착 후 use gravity 해제했습니다! 그랬는데도 똑같네요... 플레이어 태그라는게 이부분이 맞나요??
> 
> ![image](https://user-images.githubusercontent.com/105036393/167526987-badbdcb5-6bf4-4596-865f-89029e4b323c.png)

큐브 스크립트에서 OnTriggerEnter 에 console.log 로 아무거나 출력 해보시고 출력이 되는지 확인요 ~

**Reply by maskpop** [reply]
아래 상태일때는 로그가 뜹니다.
![image](https://user-images.githubusercontent.com/105036393/167530830-6b803546-c4c4-41b3-bb3d-1b7d82dc7ed7.png)

하지만 아래일 때는 뜨지 않아요...
![image](https://user-images.githubusercontent.com/105036393/167530900-f045587e-b968-418b-b301-09807fab5e60.png)

if Player 추가하면 로그가 뜨질 않습니다..

**Reply by smilejsu82** [reply]
> 아래 상태일때는 로그가 뜹니다.
> 
> ![image](https://user-images.githubusercontent.com/105036393/167530830-6b803546-c4c4-41b3-bb3d-1b7d82dc7ed7.png)
> 
> 
> 
> 하지만 아래일 때는 뜨지 않아요...
> 
> ![image](https://user-images.githubusercontent.com/105036393/167530900-f045587e-b968-418b-b301-09807fab5e60.png)
> 
> 
> 
> if Player 추가하면 로그가 뜨질 않습니다..

플레이중일때 플레이어의 tag가 Player인지 확인 되나요?

**Reply by maskpop** [reply]
플레이중에 ZepetoPlayers에서 LocalPlayer가 Player로 되어있습니다  어떤 걸 확인하면 될까요???

**Reply by maskpop** [reply]
![image](https://user-images.githubusercontent.com/105036393/167532981-85aa54af-ad4c-46fe-9f8e-c94ec37e46f5.png)

**Reply by smilejsu82** [reply]
> ![image](https://user-images.githubusercontent.com/105036393/167532981-85aa54af-ad4c-46fe-9f8e-c94ec37e46f5.png)

아 멀티 플레이세요?~

**Reply by maskpop** [reply]
아 저는 이걸 줍고 줍다보니 들어간거긴 한데 최종적으로는 제페토에 제출하려고 합니다.. 멀티플레이가 들어가면 안되는게 있나요??

**Reply by smilejsu82** [reply]
> 아 저는 이걸 줍고 줍다보니 들어간거긴 한데 최종적으로는 제페토에 제출하려고 합니다.. 멀티플레이가 들어가면 안되는게 있나요??

일단 싱글 플레이에서 동작 하게 만들고 멀티플레이 동기화 해보심이 어떨까요?

**Reply by maskpop** [reply]
해결 댔어요!! 감사합니다 친절하게 설명해주셔서 정말 좋았어요!!! 정말 감사합니다 ㅠㅠㅠ

**Reply by smilejsu82** [reply]
> 해결 댔어요!! 감사합니다 친절하게 설명해주셔서 정말 좋았어요!!! 정말 감사합니다 ㅠㅠㅠ

다행이네요 화이팅!!! 👍
